Support moving files to other package

testforstephen opened the pull request:
Signed-off-by: Jinbo Wang <jinbwan@microsoft.com>

It covers one **move** scenario of issue https://github.com/eclipse/eclipse.jdt.ls/issues/1089.
- moving file to other package, and update references.

snjeza commented:
test this please

fbricon commented:
Another pretty annoying issue is, when it works, after the file was moved, it's filled with errors. The file needs to be saved in order to be revalidated.

testforstephen commented:
> Another pretty annoying issue is, when it works, after the file was moved, it's filled with errors. The file needs to be saved in order to be revalidated.

Have a look at this issue. That's because all changes are cached in WorkingCopy of ICompilationUnit. And at the end of the refactoring action, there is a `Building Workspace` job is auto triggered, but it just built against the disk contents. 

For the moving operation, the file move changes are reflected to build job, but the cached text changes are not regarded by build job. So when the build diagnostics are received by VS Code, the user will see some package declaration and import errors in the editor.

One possible fix is to trigger save action after the move operation is finished.
@fbricon how do you think of it?

fbricon commented:
@snjeza dealt with that kind of issue before, I'll let her comment

snjeza commented:
> One possible fix is to trigger save action after the move operation is finished.

Agree.
Could you, please, rebase?

testforstephen commented:
rebased the conflicts.

> Moving files worked for standalone java projects but when I tried on a multimodule one (https://github.com/angelozerr/lsp4xml/tree/master/org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml), nothing happened.

Cannot reproduce the issue with the same lsp4xml project in my machine. However, i have added more error logs in the code, @fbricon could you try again and share the log? thanks.

testforstephen commented:
@fbricon ping...

fbricon commented:
@testforstephen see https://github.com/redhat-developer/vscode-java/pull/979#issuecomment-516494759

testforstephen commented:
thanks. Find the root cause, the PR updated.

testforstephen commented:
test this please

fbricon commented on org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/corext/refactoring/reorg/MoveCuUpdateCreator.java:
```diff
@@ -249,15 +249,15 @@ private void addImportToSourcePackageTypes(ICompilationUnit movedUnit, IProgress
 		for (int i= 0; i < referencedTypes.length; i++) {
 				IType iType= referencedTypes[i];
-				if (! iType.exists()) {
+			if (!iType.exists()) {
```
please fix the indentation in this file

fbricon commented on org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/corext/refactoring/reorg/MoveCuUpdateCreator.java:
```diff
@@ -285,12 +285,11 @@ private boolean simpleReferencesNeedNewImport(ICompilationUnit movedUnit, ICompi
 		}
 		if (referencingCu.getParent().equals(movedUnit.getParent()))
-		 {
+		{
```
opening bracket should be on the same line as the if statement

fbricon commented on org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/corrections/proposals/IProposalRelevance.java:
```diff
@@ -18,7 +18,7 @@
  * Interface defining relevance values for quick fixes/assists.
  *
- * @see org.eclipse.jdt.ui.text.java.IJavaCompletionProposal#getRelevance()
+ * see org.eclipse.jdt.ui.text.java.IJavaCompletionProposal#getRelevance()
```
please keep `@see` keyword

fbricon commented on org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/lsp/JavaProtocolExtensions.java:
```diff
@@ -97,4 +99,10 @@
 	CompletableFuture<RefactorWorkspaceEdit> getRefactorEdit(GetRefactorEditParams params);
+
+	@JsonRequest
+	CompletableFuture<PackageDestinationsResponse> getPackageDestinations(String[] documentUris);
```
I think `getMoveDestinations` is a better fit, as we might get other actions that need to return other package destinations, using different criteria

fbricon commented on org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/handlers/JDTLanguageServer.java:
```diff
@@ -875,6 +877,18 @@ public void didChangeWorkspaceFolders(DidChangeWorkspaceFoldersParams params) {
 
+	@Override
+	public CompletableFuture<PackageDestinationsResponse> getPackageDestinations(String[] documentUris) {
+		logInfo(">> java/getPackageDestinations");
```
`java/getMoveDestinations`

fbricon commented on org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/text/correction/RefactorProposalUtility.java:
```diff
@@ -50,14 +51,25 @@
+		int relevance = IProposalRelevance.MOVE_REFACTORING;
+		ICompilationUnit cu = context.getCompilationUnit();
+		String uri = JDTUtils.toURI(cu);
+		result.add(new CUCorrectionCommandProposal(label, JavaCodeActionKind.REFACTOR_MOVE, cu, relevance, RefactorProposalUtility.APPLY_REFACTORING_COMMAND_ID, Arrays.asList(MOVE_FILE_COMMAND, params, new MoveFileInfo(uri))));
```
or `return Collections.singletonList(...);`

testforstephen commented:
PR updated.

fbricon commented on org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/ChangeUtil.java:
```diff
@@ -216,29 +241,28 @@ private static void convertTextChange(TextChange textChange, WorkspaceEdit rootE
+		if (JavaLanguageServerPlugin.getPreferencesManager().getClientPreferences().isResourceOperationSupported()) {
+			List<Either<TextDocumentEdit, ResourceOperation>> changes = root.getDocumentChanges();
+			if (changes == null) {
+				changes = new LinkedList<>();
```
why not ArrayList?

testforstephen replied:
nice catch.

fbricon commented on org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/handlers/JDTLanguageServer.java:
```diff
@@ -875,6 +878,18 @@ public void didChangeWorkspaceFolders(DidChangeWorkspaceFoldersParams params) {
+	}
+
+	@Override
+	public CompletableFuture<RefactorWorkspaceEdit> moveFile(MoveFileParams params) {
```
I think we should rename it to moveResources(MoveResourcesParams params), since we might handle package moves in the future

testforstephen replied:
A pretty good suggestion, thanks!

But i prefer to make the api more generic `move(MoveParams params)`, actually we will have more move scenarios. I'm also working on support moveInstanceMethod, moveStaticMember, moveTypeToNewFile, moveTypeToOtherClass.

testforstephen commented on org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/ChangeUtil.java:
```diff
@@ -134,9 +136,32 @@ private static void convertResourceChange(ResourceChange resourceChange, Workspa
+			// update the package declaration
+			updatePackageStatement(astCU, newPackage.getElementName(), rewrite, unit);
+			convertTextEdit(edit, unit, rewrite.rewriteAST());
+		} catch (JavaModelException | IllegalArgumentException e) {
```
Here i would like to remove the try-catch block, just let the outer to deal with the exception.

testforstephen commented:
@fbricon  PR updated.

testforstephen commented:
test this please

testforstephen commented:
retest this please

fbricon commented:
I believe we now have a good starting point for implementing/improving move support. Nice job @testforstephen!

Couple things it'd be nice to address next:
- the ability to move to a new package (so currently not listed). Currently you need to create the destination package first, before executing the move command,
- the ability to undo moves (that's a client side issue though). Currently you either need to perform another move to revert the changes or, if under an SCM, revert to the previous commit

testforstephen commented:
Create two issues to track the future discussions for the improvement.
https://github.com/redhat-developer/vscode-java/issues/1017
https://github.com/redhat-developer/vscode-java/issues/1018

About undo moves, we already started the talk with VS Code team. See https://github.com/microsoft/vscode/issues/77190

GabrielBB commented:
I wonder if, later, you can make this work after you move a file by dragging

testforstephen commented:
VS Code drag and drop api is not available. If it's ready, we can enable that.
